mapper/tedge_mapper/src/sm_c8y_mapper/smartrest_serializer.rs

```diff
@@ -0,0 +1,554 @@
+        }
+    }
+
+    pub(crate) fn from_thin_edge_json(
+        response: SoftwareUpdateResponse,
+    ) -> Result<Self, SmartRestSerializerError> {
+        match response.status() {
+            SoftwareOperationStatus::Successful => {
+                Ok(Self::new(CumulocitySupportedOperations::C8ySoftwareUpdate))
+            }
+            _ => Err(SmartRestSerializerError::StatusIsNotSuccessful { response }),
+        }
+    }
+}
+
+impl<'a> Serializer<'a> for SmartRestSetOperationToSuccessful {}
+
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+pub(crate) struct SmartRestSetOperationToFailed {
+    pub message_id: &'static str,
+    pub operation: &'static str,
+    pub reason: String,
+}
+
+impl SmartRestSetOperationToFailed {
+    fn new(operation: CumulocitySupportedOperations, reason: String) -> Self {
+        Self {
+            message_id: "502",
+            operation: operation.into(),
+            reason,
+        }
+    }
+
+    pub(crate) fn from_thin_edge_json(
+        response: SoftwareUpdateResponse,
+    ) -> Result<Self, SmartRestSerializerError> {
+        match &response.status() {
+            SoftwareOperationStatus::Failed => Ok(Self::new(
+                CumulocitySupportedOperations::C8ySoftwareUpdate,
+                response.error().unwrap_or_else(|| "".to_string()),
+            )),
+            _ => Err(SmartRestSerializerError::StatusIsNotFailed { response }),
+        }
+    }
+}
+
+impl<'a> Serializer<'a> for SmartRestSetOperationToFailed {
+    // If we add double quotes before serialization, csv crate adds two double quotes to escape.
+    // Therefore, add double quotes after serialization. But I don't like this implementation.
+    fn to_smartrest(&self) -> Result<SmartRest, SmartRestSerializerError> {
+        let mut s = serialize_smartrest(self)?;
+        let position_second_comma = find_second_comma(s.as_str())?;
+        let _ = s.insert(position_second_comma + 1, '"'); // insert a double-quote after the second comma
+        let _ = s.insert(s.len() - 1, '"'); // insert a double-quote before \n
+        Ok(s)
+    }
+}
+
+fn serialize_smartrest<S: Serialize>(record: S) -> Result<String, SmartRestSerializerError> {
+    let mut wtr = WriterBuilder::new()
+        .has_headers(false)
+        .quote_style(QuoteStyle::Never)
+        .double_quote(false)
+        .from_writer(vec![]);
+    wtr.serialize(record)?;
+    let csv = String::from_utf8(wtr.into_inner()?)?;
+    Ok(csv)
+}
+
+fn find_second_comma(s: &str) -> Result<usize, SmartRestSerializerError> {
+    let mut s_clone = s.to_string();
+    let position_first_comma = s_clone
+        .find(',')
+        .ok_or(SmartRestSerializerError::DoubleQuoteError)?;
+    let s_after_comma = s_clone.split_off(position_first_comma + 1);
+    let position_second_comma = s_after_comma
+        .find(',')
+        .ok_or(SmartRestSerializerError::DoubleQuoteError)?;
+    Ok(position_first_comma + position_second_comma + 1)
+}
```

Comment:
I think if you use `serde` as per above comment you may not need those functions anymore.